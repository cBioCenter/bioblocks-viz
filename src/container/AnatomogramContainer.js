"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
// tslint:disable-next-line:import-name
var anatomogram_1 = require("anatomogram");
var immutable_1 = require("immutable");
var React = require("react");
var react_redux_1 = require("react-redux");
var redux_1 = require("redux");
var action_1 = require("~bioblocks-viz~/action");
var component_1 = require("~bioblocks-viz~/component");
var container_1 = require("~bioblocks-viz~/container");
var data_1 = require("~bioblocks-viz~/data");
var helper_1 = require("~bioblocks-viz~/helper");
var reducer_1 = require("~bioblocks-viz~/reducer");
var selector_1 = require("~bioblocks-viz~/selector");
var AnatomogramContainerClass = /** @class */ (function (_super) {
    tslib_1.__extends(AnatomogramContainerClass, _super);
    function AnatomogramContainerClass(props) {
        var _this = _super.call(this, props) || this;
        _this.divRef = null;
        _this.onClick = function (ids) {
            // Anatomogram returns an array of strings for click events, but we only ever work with a single id.
            var id = ids[0];
            var _a = _this.props, addLabel = _a.onLabelAdd, removeLabel = _a.onLabelRemove, selectIds = _a.selectIds;
            if (selectIds.includes(id)) {
                removeLabel(id);
            }
            else {
                addLabel(id);
            }
        };
        _this.deriveIdsFromSpecies = function (species) { return Object.keys(data_1.AnatomogramMapping[species]); };
        _this.onMouseOut = function (id) {
            return;
        };
        _this.onMouseOver = function (id) {
            return;
        };
        _this.parseCategory = function (category) {
            var splitCategories = category.split(/-|_/);
            return splitCategories[0];
        };
        _this.resizeSVGElement = function (error, svgDomNode) {
            if (_this.divRef) {
                var isSvgHeightBigger = svgDomNode.height.baseVal.value > svgDomNode.width.baseVal.value;
                // The Anatomogram Component internally sets the svg height to 'auto'.
                // So, to allow more flexibility in sizing it, we have to manually override it here. Sorry.
                svgDomNode.style.height = isSvgHeightBigger ? "calc(" + _this.divRef.style.height + " - 50px)" : 'auto';
                svgDomNode.style.padding = '0';
                svgDomNode.style.width = isSvgHeightBigger ? 'auto' : "calc(" + _this.divRef.style.width + " - 75px)";
            }
        };
        _this.state = {
            ids: _this.deriveIdsFromSpecies(props.species),
        };
        return _this;
    }
    AnatomogramContainerClass.prototype.setupDataServices = function () {
        this.registerDataset('cells', []);
        this.registerDataset('labels', []);
        reducer_1.BioblocksMiddlewareTransformer.addTransform(this.getCellToLabelTransform());
        reducer_1.BioblocksMiddlewareTransformer.addTransform(this.getLabelToCellTransform());
    };
    AnatomogramContainerClass.prototype.componentDidUpdate = function (prevProps) {
        var species = this.props.species;
        if (species !== prevProps.species) {
            this.setState({
                ids: this.deriveIdsFromSpecies(species),
            });
        }
    };
    AnatomogramContainerClass.prototype.render = function () {
        var _this = this;
        var _a = this.props, iconSrc = _a.iconSrc, species = _a.species, selectIds = _a.selectIds;
        var ids = this.state.ids;
        return (React.createElement(react_redux_1.Provider, { store: reducer_1.BBStore },
            React.createElement("div", { className: 'anatomogram-container', ref: function (node) {
                    if (node) {
                        _this.divRef = node.getElementsByTagName('div')[0];
                    }
                } },
                React.createElement(component_1.ComponentCard, { componentName: 'Anatomogram', iconSrc: iconSrc },
                    React.createElement(anatomogram_1.default, { atlasUrl: "", highlightColour: 'yellow', onClick: this.onClick, onInjected: this.resizeSVGElement, onMouseOut: this.onMouseOut, onMouseOver: this.onMouseOver, selectColour: 'ffaa00', selectIds: selectIds.toArray(), showIds: ids, species: species, selectedView: species === 'mus_musculus' ? 'female' : 'male' })))));
    };
    AnatomogramContainerClass.prototype.getCellToLabelTransform = function () {
        var _this = this;
        return {
            fn: function (state) {
                var e_1, _a, e_2, _b, e_3, _c;
                var species = _this.props.species;
                var currentCells = selector_1.selectCurrentItems(state, 'cells').toArray();
                var _d = selector_1.getSpring(state), category = _d.category, graphData = _d.graphData;
                var result = immutable_1.Set();
                try {
                    for (var currentCells_1 = tslib_1.__values(currentCells), currentCells_1_1 = currentCells_1.next(); !currentCells_1_1.done; currentCells_1_1 = currentCells_1.next()) {
                        var cellIndex = currentCells_1_1.value;
                        // Parse labels and categories from SPRING data.
                        var labelForCategory = graphData.nodes[cellIndex] ? graphData.nodes[cellIndex].labelForCategory : {};
                        var labels = data_1.AnatomogramMapping[species][labelForCategory[category]];
                        if (labels) {
                            labels.forEach(function (label) { return (result = result.add(label)); });
                        }
                        try {
                            for (var _e = (e_2 = void 0, tslib_1.__values(Object.keys(data_1.AnatomogramMapping[species]))), _f = _e.next(); !_f.done; _f = _e.next()) {
                                var id = _f.value;
                                try {
                                    for (var _g = (e_3 = void 0, tslib_1.__values(Object.values(labelForCategory))), _h = _g.next(); !_h.done; _h = _g.next()) {
                                        var label = _h.value;
                                        if (data_1.AnatomogramMapping[species][id].includes(label)) {
                                            result = result.add(id);
                                        }
                                    }
                                }
                                catch (e_3_1) { e_3 = { error: e_3_1 }; }
                                finally {
                                    try {
                                        if (_h && !_h.done && (_c = _g.return)) _c.call(_g);
                                    }
                                    finally { if (e_3) throw e_3.error; }
                                }
                            }
                        }
                        catch (e_2_1) { e_2 = { error: e_2_1 }; }
                        finally {
                            try {
                                if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                            }
                            finally { if (e_2) throw e_2.error; }
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (currentCells_1_1 && !currentCells_1_1.done && (_a = currentCells_1.return)) _a.call(currentCells_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
                return result;
            },
            fromState: { stateName: 'cells' },
            toState: { stateName: 'labels' },
        };
    };
    AnatomogramContainerClass.prototype.getLabelToCellTransform = function () {
        var _this = this;
        return {
            fn: function (state) {
                var _a = _this.props, species = _a.species, selectIds = _a.selectIds;
                var graphData = selector_1.getSpring(state).graphData;
                var anatomogramMap = data_1.AnatomogramMapping[species];
                var candidateLabels = immutable_1.Set();
                selectIds.forEach(function (id) {
                    candidateLabels = id && anatomogramMap[id] ? candidateLabels.merge(anatomogramMap[id]) : candidateLabels;
                });
                var cellIndices = immutable_1.Set();
                graphData.nodes.forEach(function (node) {
                    var labelsForNode = Object.values(node.labelForCategory);
                    candidateLabels.forEach(function (label) {
                        if (label && labelsForNode.includes(label)) {
                            cellIndices = cellIndices.add(node.number);
                            return;
                        }
                    });
                });
                return cellIndices;
            },
            fromState: 'bioblocks/labels',
            toState: 'bioblocks/cells',
        };
    };
    AnatomogramContainerClass.defaultProps = {
        iconSrc: '',
        onLabelAdd: helper_1.EMPTY_FUNCTION,
        onLabelRemove: helper_1.EMPTY_FUNCTION,
        selectIds: immutable_1.Set(),
        species: 'homo_sapiens',
    };
    AnatomogramContainerClass.displayName = 'Anatomogram';
    return AnatomogramContainerClass;
}(container_1.BioblocksVisualization));
exports.AnatomogramContainerClass = AnatomogramContainerClass;
var mapStateToProps = function (state) { return ({
    selectIds: selector_1.selectCurrentItems(state, 'labels'),
}); };
var mapDispatchToProps = function (dispatch) {
    return redux_1.bindActionCreators({
        onLabelAdd: action_1.createContainerActions('labels').add,
        onLabelRemove: action_1.createContainerActions('labels').remove,
    }, dispatch);
};
exports.AnatomogramContainer = react_redux_1.connect(mapStateToProps, mapDispatchToProps)(AnatomogramContainerClass);
//# sourceMappingURL=AnatomogramContainer.js.map