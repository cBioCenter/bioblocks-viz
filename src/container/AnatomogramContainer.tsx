// tslint:disable-next-line:import-name
import Anatomogram from 'anatomogram';
import { isEqual } from 'lodash';
import * as React from 'react';

import {
  CellContext,
  ICellContext,
  initialCellContext,
  initialSpringContext,
  ISpringContext,
  SpringContext,
} from '~chell-viz~/context';
import { CHELL_CSS_STYLE } from '~chell-viz~/data';

export interface IAnatomogramContainerProps {
  cellContext: ICellContext;
  height: number | string;
  species: 'homo_sapiens' | 'mus_musculus';
  springContext: ISpringContext;
  style: CHELL_CSS_STYLE;
  width: number | string;
}

export interface IAnatomogramContainerState {
  ids: string[];
  selectIds: string[];
}

export const springToAnatomogramMapping: { [key: string]: { [key: string]: string } } = {
  homo_sapiens: {
    P11A: 'UBERON_0000178',
    P11B: 'UBERON_0000473',
    P12A: 'UBERON_0000955',
    P9A: 'UBERON_0000977',
  },
  mus_musculus: {
    bladder: 'UBERON_0001255',
    heart: 'UBERON_0000948',
    kidney: 'UBERON_0002113',
    liver: 'UBERON_0002107',
    lung: 'UBERON_0002048 ',
    mammary: 'UBERON_0001911',
    marrow: 'UBERON_0002371',
    muscle: 'sUBERON_0014892',
    spleen: 'UBERON_0002106',
    thymus: 'UBERON_0002370',
    tongue: 'UBERON_0001723',
    trachea: 'UBERON_0003126',
  },
};

export const anatomogramToSpringMapping: { [key: string]: { [key: string]: string } } = {
  homo_sapiens: {
    UBERON_0000178: 'P11A',
    UBERON_0000473: 'P11B',
    UBERON_0000955: 'P12A',
    UBERON_0000977: 'P9A',
  },
  mus_musculus: {
    UBERON_0000948: 'heart',
    UBERON_0001255: 'bladder',
    UBERON_0001723: 'tongue',
    UBERON_0001911: 'mammary',
    UBERON_0002048: 'lung',
    UBERON_0002106: 'spleen',
    UBERON_0002107: 'liver',
    UBERON_0002113: 'kidney',
    UBERON_0002370: 'thymus',
    UBERON_0002371: 'marrow',
    UBERON_0003126: 'trachea',
    sUBERON_0014892: 'muscle',
  },
};

export class AnatomogramContainerClass extends React.Component<IAnatomogramContainerProps, IAnatomogramContainerState> {
  public static defaultProps = {
    cellContext: {
      ...initialCellContext,
    },
    height: '300px',
    ids: ['UBERON_0000178', 'UBERON_0000473', 'UBERON_0000955', 'UBERON_0000977'],
    springContext: {
      ...initialSpringContext,
    },
    style: {},
    width: '400px',
  };

  public static displayName = 'Anatomogram';

  protected anatomogramRef: any;

  constructor(props: IAnatomogramContainerProps) {
    super(props);
    this.state = {
      ids: Object.keys(anatomogramToSpringMapping[props.species]),
      selectIds: [],
    };
  }

  public componentDidUpdate(prevProps: IAnatomogramContainerProps) {
    const { cellContext, springContext } = this.props;
    if (!isEqual(springContext.selectedCategories, prevProps.springContext.selectedCategories)) {
      this.setState({
        selectIds: springContext.selectedCategories.map(
          category => springToAnatomogramMapping[this.props.species][category],
        ),
      });
    } else if (!isEqual(cellContext.currentCells, prevProps.cellContext.currentCells)) {
      const categories = new Set<string>();
      for (const cellIndex of cellContext.currentCells) {
        categories.add(springContext.graphData.nodes[cellIndex].category);
      }
      this.setState({
        selectIds: Array.from(categories).map(category => springToAnatomogramMapping[this.props.species][category]),
      });
    }
  }

  public render() {
    const { ids, selectIds } = this.state;

    return (
      <div className={'anatomogram-container'} style={{ height: '100%' }}>
        <Anatomogram
          atlasUrl={``}
          highlightColour={'yellow'}
          onClick={this.onClick}
          onMouseOut={this.onMouseOut}
          onMouseOver={this.onMouseOver}
          selectColour={'ffaa00'}
          selectIds={selectIds}
          showIds={ids}
          species={this.props.species}
          selectedView={this.props.species === 'mus_musculus' ? 'female' : 'male'}
        />
      </div>
    );
  }

  protected onClick = (id: string) => {
    this.props.springContext.toggleCategory(anatomogramToSpringMapping[this.props.species][id]);
  };

  protected onMouseOut = (id: string) => {
    return;
  };

  protected onMouseOver = (id: string) => {
    return;
  };
}

type requiredProps = Omit<IAnatomogramContainerProps, keyof typeof AnatomogramContainerClass.defaultProps> &
  Partial<IAnatomogramContainerProps>;

export const AnatomogramContainer = (props: requiredProps) => (
  <CellContext.Consumer>
    {cellContext => (
      <SpringContext.Consumer>
        {springContext => (
          <AnatomogramContainerClass {...props} cellContext={cellContext} springContext={springContext} />
        )}
      </SpringContext.Consumer>
    )}
  </CellContext.Consumer>
);
