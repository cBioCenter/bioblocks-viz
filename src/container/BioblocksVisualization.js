"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var immutable_1 = require("immutable");
var React = require("react");
var util_1 = require("util");
var data_1 = require("~bioblocks-viz~/data");
var reducer_1 = require("~bioblocks-viz~/reducer");
var BioblocksVisualization = /** @class */ (function (_super) {
    tslib_1.__extends(BioblocksVisualization, _super);
    function BioblocksVisualization(props) {
        var _this = _super.call(this, props) || this;
        _this.bioblocksHooks = {};
        _this.datasets = immutable_1.Set();
        _this.loadingStatus = 2 /* NOT_STARTED */;
        BioblocksVisualization.activeBioblocksVisualizations = BioblocksVisualization.activeBioblocksVisualizations.add(_this);
        _this.setupDataServices();
        return _this;
    }
    BioblocksVisualization.prototype.componentWillUnmount = function () {
        this.teardown();
    };
    BioblocksVisualization.prototype.addBioblocksHook = function (name, cb) {
        this.bioblocksHooks[name] = cb;
    };
    BioblocksVisualization.prototype.registerDataset = function (name, defaultValue, namespace) {
        if (namespace === void 0) { namespace = 'bioblocks'; }
        this.datasets = this.datasets.add(new data_1.Dataset(name, namespace));
        this.createReducer(name, defaultValue, namespace);
    };
    BioblocksVisualization.prototype.finishLoading = function () {
        this.loadingStatus = 3 /* READY */;
    };
    BioblocksVisualization.prototype.getDatasets = function () {
        return this.datasets;
    };
    BioblocksVisualization.prototype.getComponentBioblocksHooks = function () {
        return this.bioblocksHooks;
    };
    BioblocksVisualization.prototype.getLoadingStatus = function () {
        return this.loadingStatus;
    };
    BioblocksVisualization.prototype.teardown = function () {
        // TODO Can it be enforced that this method is always called?
        BioblocksVisualization.activeBioblocksVisualizations = BioblocksVisualization.activeBioblocksVisualizations.remove(this);
    };
    BioblocksVisualization.prototype.createReducer = function (datasetName, defaultValue, namespace) {
        if (namespace === void 0) { namespace = 'bioblocks'; }
        if (util_1.isArray(defaultValue)) {
            reducer_1.createContainerReducer(datasetName, namespace);
        }
        else if (typeof defaultValue === 'object') {
            reducer_1.createObjectReducer(datasetName, namespace);
        }
        else {
            reducer_1.createValueReducer(datasetName, defaultValue, namespace);
        }
    };
    BioblocksVisualization.getActiveBioblocksVisualizations = function () {
        return BioblocksVisualization.activeBioblocksVisualizations;
    };
    BioblocksVisualization.activeBioblocksVisualizations = immutable_1.Set();
    return BioblocksVisualization;
}(React.Component));
exports.BioblocksVisualization = BioblocksVisualization;
//# sourceMappingURL=BioblocksVisualization.js.map