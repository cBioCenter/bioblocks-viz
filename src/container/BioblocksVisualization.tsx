import { Set } from 'immutable';
import * as React from 'react';
import { isArray } from 'util';

import { Dataset } from '~bioblocks-viz~/data';
import {
  createContainerReducer,
  createDataReducer,
  createObjectReducer,
  createValueReducer,
} from '~bioblocks-viz~/reducer';

export const enum BIOBLOCKS_LOADING_STATUS {
  'ERROR',
  'LOADING',
  'NOT_STARTED',
  'READY',
}

export type BioblocksHookCallback<T> = () => T;

export interface IBioblocksHookDict {
  [key: string]: BioblocksHookCallback<any>;
}

export abstract class BioblocksVisualization<P = any, S = any, SS = any> extends React.Component<P, S, SS> {
  public static getActiveBioblocksVisualizations = () => {
    return BioblocksVisualization.activeBioblocksVisualizations;
  };

  public static getActiveBioblocksHooks = () => {
    let allHooks: IBioblocksHookDict = {};

    BioblocksVisualization.activeBioblocksVisualizations.forEach(viz => {
      if (viz) {
        allHooks = {
          ...allHooks,
          ...viz.bioblocksHooks,
        };
      }
    });

    return allHooks;
  };

  public static getActiveDatasets = () => {
    let allDatasets = Set<Dataset>();

    BioblocksVisualization.activeBioblocksVisualizations.forEach(viz => {
      if (viz) {
        allDatasets = allDatasets.merge(viz.datasets);
      }
    });

    return allDatasets;
  };

  private static activeBioblocksVisualizations = Set<BioblocksVisualization>();

  protected bioblocksHooks: IBioblocksHookDict = {};
  protected datasets = Set<Dataset>();

  private loadingStatus = BIOBLOCKS_LOADING_STATUS.NOT_STARTED;

  constructor(props: P) {
    super(props);
    BioblocksVisualization.activeBioblocksVisualizations = BioblocksVisualization.activeBioblocksVisualizations.add(
      this,
    );
    this.setupDataServices();
  }

  public abstract setupDataServices(): void;

  public componentWillUnmount() {
    this.teardown();
  }

  public addBioblocksHook<T>(name: string, cb: () => T) {
    this.bioblocksHooks[name] = cb;
  }

  public registerDataset<T>(name: string, defaultValue?: T, namespace = 'bioblocks') {
    this.datasets = this.datasets.add(new Dataset(name, namespace));
    this.createReducer(name, defaultValue, namespace);
  }

  public finishLoading() {
    this.loadingStatus = BIOBLOCKS_LOADING_STATUS.READY;
  }

  public getDatasets() {
    return this.datasets;
  }

  public getComponentBioblocksHooks() {
    return this.bioblocksHooks;
  }

  public getLoadingStatus() {
    return this.loadingStatus;
  }

  public teardown() {
    // TODO Can it be enforced that this method is always called?
    BioblocksVisualization.activeBioblocksVisualizations = BioblocksVisualization.activeBioblocksVisualizations.remove(
      this,
    );
  }

  protected createReducer<T>(datasetName: string, defaultValue?: T, namespace = 'bioblocks') {
    if (isArray(defaultValue)) {
      createContainerReducer<T>(datasetName, namespace);
    } else if (typeof defaultValue === 'object') {
      createObjectReducer<T>(datasetName, namespace);
    } else if (defaultValue === undefined) {
      createDataReducer<T>(datasetName, namespace);
    } else {
      createValueReducer<T>(datasetName, namespace);
    }
  }
}
