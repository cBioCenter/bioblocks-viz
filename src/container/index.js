"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
tslib_1.__exportStar(require("./BioblocksVisualization"), exports);
tslib_1.__exportStar(require("./AnatomogramContainer"), exports);
tslib_1.__exportStar(require("./InfoPanelContainer"), exports);
tslib_1.__exportStar(require("./NGLContainer"), exports);
tslib_1.__exportStar(require("./ContactMapContainer"), exports);
tslib_1.__exportStar(require("./ProteinFeatureViewer"), exports);
tslib_1.__exportStar(require("./SpringContainer"), exports);
tslib_1.__exportStar(require("./TensorTContainer"), exports);
tslib_1.__exportStar(require("./UMAPSequenceContainer"), exports);
tslib_1.__exportStar(require("./UMAPTranscriptionalContainer"), exports);
//# sourceMappingURL=index.js.map