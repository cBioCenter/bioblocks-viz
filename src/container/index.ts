export * from './BioblocksVisualization';

export * from './AnatomogramContainer';
export * from './InfoPanelContainer';
export * from './NGLContainer';
export * from './ContactMapContainer';
export * from './ProteinFeatureViewer';
export * from './SpringContainer';
export * from './TensorTContainer';
export * from './UMAPSequenceContainer';
export * from './UMAPTranscriptionalContainer';
