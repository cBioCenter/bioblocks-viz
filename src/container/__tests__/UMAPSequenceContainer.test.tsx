import { mount, shallow } from 'enzyme';
import * as React from 'react';
import { act } from 'react-test-renderer';

import { UMAPSequenceContainerClass } from '~bioblocks-viz~/container';
import { Seq, SeqRecord } from '~bioblocks-viz~/data';
import { flushPromises } from '~bioblocks-viz~/test';

// tslint:disable-next-line: max-func-body-length
describe('UMAPSequenceContainer', () => {
  let sequence: Seq;
  let sequences: SeqRecord[];
  let taxonomyText: string;
  beforeEach(() => {
    jest.clearAllTimers();
    jest.useFakeTimers();
    sequence = new Seq('gatcctag');
    sequences = [new SeqRecord(sequence)];
    // Taxonomy text based off betalactamase_alignment - PSE1_NATURAL_TAXONOMY.csv
    taxonomyText = `seq_name,sequence,tax_id,superkingdom,phylum,genus,class,subphylum,family,order,species\n\
example_sequence,EVA-NERV,59846,Bacteria,Firmicutes,Paenibacillus,Bacilli,Paenibacillaceae,Bacillales,Paenibacillus,chibensis`;
  });

  it('Should render when given an empty sequence.', () => {
    const wrapper = shallow(<UMAPSequenceContainerClass allSequences={[]} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('Should render when given a sequence.', () => {
    const wrapper = shallow(<UMAPSequenceContainerClass allSequences={sequences} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('Should handle a taxonomy update.', () => {
    const wrapper = shallow(<UMAPSequenceContainerClass allSequences={sequences} />);
    const expected = {
      example_sequence: {
        class: 'Bacilli',
        family: 'Bacillales',
        genus: 'Paenibacillus',
        order: 'Paenibacillus',
        phylum: 'Firmicutes',
        seq_name: 'example_sequence',
        sequence: 'EVA-NERV',
        species: 'chibensis',
        subphylum: 'Paenibacillaceae',
        superkingdom: 'Bacteria',
        tax_id: '59846',
      },
    };
    wrapper.setProps({
      taxonomyText,
    });
    expect(wrapper.state('seqNameToTaxonomyMetadata')).toEqual(expected);
  });

  it('Should render an upload form when enabled.', () => {
    const wrapper = shallow(<UMAPSequenceContainerClass allSequences={sequences} showUploadButton={true} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('Should parse a taxonomy file with headers.', async () => {
    const wrapper = mount(
      <UMAPSequenceContainerClass allSequences={[]} labelCategory={'example_sequence'} showUploadButton={true} />,
    );
    const instance = wrapper.instance() as UMAPSequenceContainerClass;
    expect(instance.state.labels).toEqual([]);

    await act(async () => {
      wrapper
        .find('input')
        .simulate('change', { target: { files: { item: () => new File([taxonomyText], 'mock'), length: 1 } } });
    });

    await act(async () => {
      wrapper.update();
      instance.forceUpdate();
    });

    const expected = [
      'seq_name',
      'tax_id',
      'superkingdom',
      'phylum',
      'genus',
      'class',
      'subphylum',
      'family',
      'order',
      'species',
    ];
    expect(instance.state.labels).toEqual(expected);
  });

  it('Should parse a taxonomy file without headers.', async () => {
    const wrapper = mount(
      <UMAPSequenceContainerClass allSequences={[]} labelCategory={'example_sequence'} showUploadButton={true} />,
    );
    const instance = wrapper.instance() as UMAPSequenceContainerClass;
    expect(instance.state.labels).toEqual([]);
    const taxonomyTextNoHeaders = taxonomyText.split('\n')[1];
    await act(async () => {
      wrapper.find('input').simulate('change', {
        target: { files: { item: () => new File([taxonomyTextNoHeaders], 'mock'), length: 1 } },
      });
    });

    await act(async () => {
      wrapper.update();
    });

    const expected = new Array();
    expect(instance.state.labels).toEqual(expected);
  });

  it('Should gracefully handle an invalid taxonomy file.', async () => {
    const wrapper = mount(
      <UMAPSequenceContainerClass allSequences={[]} labelCategory={'example_sequence'} showUploadButton={true} />,
    );
    const instance = wrapper.instance() as UMAPSequenceContainerClass;
    expect(instance.state.labels).toEqual([]);
    await act(async () => {
      wrapper.find('input').simulate('change', { target: { files: null } });
    });

    await act(async () => {
      wrapper.update();
    });

    expect(instance.state.labels).toEqual([]);
  });

  it('Should handle updating the data.', async () => {
    const wrapper = shallow(<UMAPSequenceContainerClass allSequences={sequences} />);
    wrapper.setProps({
      allSequences: [new SeqRecord(new Seq('ggaattcc'))],
    });
    await act(async () => {
      wrapper.update();
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('Should handle removing the taxonomy data.', async () => {
    const wrapper = shallow(<UMAPSequenceContainerClass allSequences={sequences} taxonomyText={taxonomyText} />);
    wrapper.setProps({
      taxonomyText: undefined,
    });
    await act(async () => {
      wrapper.update();
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('Should handle missing annotation data.', async () => {
    const unannotatedSequences = sequences.map(seq => {
      seq.annotations.name = undefined;

      return seq;
    });

    const wrapper = shallow(
      <UMAPSequenceContainerClass allSequences={unannotatedSequences} taxonomyText={taxonomyText} />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('Should use hamming method for distance calculation.', async () => {
    const wrapper = mount(
      <UMAPSequenceContainerClass
        allSequences={[
          ...sequences,
          ...sequences,
          ...sequences,
          ...sequences,
          ...sequences,
          ...sequences,
          ...sequences,
          new SeqRecord(new Seq('catCat')),
          ...sequences,
          ...sequences,
          ...sequences,
          ...sequences,
          ...sequences,
          ...sequences,
          ...sequences,
          ...sequences,
        ]}
      />,
    );
    const instance = wrapper.instance() as UMAPSequenceContainerClass;
    const spy = jest.spyOn(instance, 'equalityHammingDistance');
    expect(spy).not.toHaveBeenCalled();
    await act(async () => {
      wrapper.update();
      instance.forceUpdate();
    });
    jest.runAllTimers();
    expect(instance.equalityHammingDistance).toHaveBeenCalled();
  });
});
