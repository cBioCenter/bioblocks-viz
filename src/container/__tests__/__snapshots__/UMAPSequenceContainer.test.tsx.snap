// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UMAPSequenceContainer Should handle missing annotation data. 1`] = `
<Grid
  centered={true}
  padded={true}
>
  <GridRow>
    <UMAPVisualization
      currentLabel="class"
      dataLabels={
        Array [
          undefined,
        ]
      }
      dataMatrix={
        Array [
          Array [
            0,
            1,
            2,
            3,
            3,
            2,
            1,
            0,
          ],
        ]
      }
      distanceFn={[Function]}
      errorMessages={Array []}
      labelCategory="class"
      labels={
        Array [
          "seq_name",
          "tax_id",
          "superkingdom",
          "phylum",
          "genus",
          "class",
          "subphylum",
          "family",
          "order",
          "species",
        ]
      }
      minDist={0.99}
      nComponents={2}
      nNeighbors={15}
      numIterationsBeforeReRender={1}
      numSequencesToShow={4000}
      onLabelChange={[Function]}
      spread={1}
      taxonomyText="seq_name,sequence,tax_id,superkingdom,phylum,genus,class,subphylum,family,order,species
example_sequence,EVA-NERV,59846,Bacteria,Firmicutes,Paenibacillus,Bacilli,Paenibacillaceae,Bacillales,Paenibacillus,chibensis"
      tooltipNames={
        Array [
          "",
        ]
      }
    />
  </GridRow>
</Grid>
`;

exports[`UMAPSequenceContainer Should handle removing the taxonomy data. 1`] = `
<Grid
  centered={true}
  padded={true}
>
  <GridRow>
    <UMAPVisualization
      currentLabel="class"
      dataLabels={
        Array [
          undefined,
        ]
      }
      dataMatrix={
        Array [
          Array [
            0,
            1,
            2,
            3,
            3,
            2,
            1,
            0,
          ],
        ]
      }
      distanceFn={[Function]}
      errorMessages={Array []}
      labelCategory="class"
      labels={
        Array [
          "seq_name",
          "tax_id",
          "superkingdom",
          "phylum",
          "genus",
          "class",
          "subphylum",
          "family",
          "order",
          "species",
        ]
      }
      minDist={0.99}
      nComponents={2}
      nNeighbors={15}
      numIterationsBeforeReRender={1}
      numSequencesToShow={4000}
      onLabelChange={[Function]}
      spread={1}
      tooltipNames={
        Array [
          "<unknown name>",
        ]
      }
    />
  </GridRow>
</Grid>
`;

exports[`UMAPSequenceContainer Should handle updating the data. 1`] = `
<Grid
  centered={true}
  padded={true}
>
  <GridRow>
    <UMAPVisualization
      currentLabel="class"
      dataLabels={
        Array [
          undefined,
        ]
      }
      dataMatrix={
        Array [
          Array [
            0,
            0,
            1,
            1,
            2,
            2,
            3,
            3,
          ],
        ]
      }
      distanceFn={[Function]}
      errorMessages={Array []}
      labelCategory="class"
      labels={Array []}
      minDist={0.99}
      nComponents={2}
      nNeighbors={15}
      numIterationsBeforeReRender={1}
      numSequencesToShow={4000}
      onLabelChange={[Function]}
      spread={1}
      tooltipNames={
        Array [
          "<unknown name>",
        ]
      }
    />
  </GridRow>
</Grid>
`;

exports[`UMAPSequenceContainer Should render an upload form when enabled. 1`] = `
<Grid
  centered={true}
  padded={true}
>
  <GridRow>
    <UMAPVisualization
      currentLabel="class"
      dataLabels={
        Array [
          undefined,
        ]
      }
      dataMatrix={
        Array [
          Array [
            0,
            1,
            2,
            3,
            3,
            2,
            1,
            0,
          ],
        ]
      }
      distanceFn={[Function]}
      errorMessages={Array []}
      labelCategory="class"
      labels={Array []}
      minDist={0.99}
      nComponents={2}
      nNeighbors={15}
      numIterationsBeforeReRender={1}
      numSequencesToShow={4000}
      onLabelChange={[Function]}
      spread={1}
      tooltipNames={
        Array [
          "<unknown name>",
        ]
      }
    />
  </GridRow>
  <GridRow>
    <Popup
      content={
        <React.Fragment>
          <span>
            A tab delimited file with 2 columns and optional headers:
          </span>
          <br />
          <span>
            - First column is sequence name.
          </span>
          <br />
          <span>
            - Second column is how to group sequences.
          </span>
        </React.Fragment>
      }
      disabled={false}
      offset={0}
      on="hover"
      pinned={false}
      position="top left"
      trigger={
        <Label
          as="label"
          basic={true}
          htmlFor="upload-taxonomy"
        >
          <Button
            as="button"
            icon="upload"
            label={
              Object {
                "basic": true,
                "content": "Upload Taxonomy File",
              }
            }
            labelPosition="right"
          />
          <input
            hidden={true}
            id="upload-taxonomy"
            multiple={false}
            onChange={[Function]}
            required={true}
            type="file"
          />
        </Label>
      }
    />
  </GridRow>
</Grid>
`;

exports[`UMAPSequenceContainer Should render when given a sequence. 1`] = `
<Grid
  centered={true}
  padded={true}
>
  <GridRow>
    <UMAPVisualization
      currentLabel="class"
      dataLabels={
        Array [
          undefined,
        ]
      }
      dataMatrix={
        Array [
          Array [
            0,
            1,
            2,
            3,
            3,
            2,
            1,
            0,
          ],
        ]
      }
      distanceFn={[Function]}
      errorMessages={Array []}
      labelCategory="class"
      labels={Array []}
      minDist={0.99}
      nComponents={2}
      nNeighbors={15}
      numIterationsBeforeReRender={1}
      numSequencesToShow={4000}
      onLabelChange={[Function]}
      spread={1}
      tooltipNames={
        Array [
          "<unknown name>",
        ]
      }
    />
  </GridRow>
</Grid>
`;

exports[`UMAPSequenceContainer Should render when given an empty sequence. 1`] = `
<Grid
  centered={true}
  padded={true}
>
  <GridRow>
    <UMAPVisualization
      currentLabel="class"
      dataLabels={Array []}
      dataMatrix={Array []}
      distanceFn={[Function]}
      errorMessages={Array []}
      labelCategory="class"
      labels={Array []}
      minDist={0.99}
      nComponents={2}
      nNeighbors={15}
      numIterationsBeforeReRender={1}
      numSequencesToShow={4000}
      onLabelChange={[Function]}
      spread={1}
      tooltipNames={Array []}
    />
  </GridRow>
</Grid>
`;
